### DISCUSSION TITLE
Dealing with Categorical Variables in Clustering

### DISCUSSION THREAD

**Discussion by twitte01** [twitte01/232R_GroupProject#68] [Model/ Algorithm Ideas]
# One-Hot Encoding
One way to handle categorical variables is to use one-hot encoding. One-hot encoding transforms categorical variables into a set of binary features, where each feature represents a distinct category. For example, suppose we have a categorical variable “color” that can take on the values red, blue, or yellow. We can transform this variable into three binary features, “color-red,” “color-blue,” and “color-yellow,” which can only take on the values 1 or 0. This increases the dimensionality of the space, but it allows us to use any clustering algorithm we like.

It is important to note that one-hot encoding is only suitable for nominal data, which does not have an inherent order. For ordinal data, such as “bad,” “average,” and “good,” it may be more appropriate to use a numerical encoding, such as 0, 1, and 2, respectively.

#### Perform one-hot encoding
encoder = OneHotEncoder(sparse=False)
one_hot_encoded = encoder.fit_transform(df[['color']])

####Create a new dataframe with one-hot encoded features
one_hot_df = pd.DataFrame(one_hot_encoded, columns=encoder.get_feature_names_out(['color']))

#### Display the one-hot encoded dataframe
print("\nOne-Hot Encoded DataFrame:")
print(one_hot_df)

**Comment by twitte01** [comment]
# K-Modes Clustering
K-means clustering is a popular clustering algorithm, but it is not directly applicable to categorical data. K-modes is a variation of k-means clustering that is specifically designed to handle categorical data. K-modes replaces the Euclidean distance metric used in k-means with a distance metric that is suitable for categorical data. K-modes works by identifying the modes (i.e., the most frequently occurring values) of the categorical variables and clustering the data points based on the mode values.

# Mixed Clustering
Mixed clustering is a technique that can handle datasets that contain both numerical and categorical variables. One way to perform mixed clustering is to use the k-prototypes algorithm, which combines k-means clustering for numerical data with k-modes clustering for categorical data. The k-prototypes algorithm uses a distance metric that combines the Euclidean distance for numerical data and the distance metric used in k-modes for categorical data.
<img width="686" alt="Screenshot 2024-05-16 at 9 22 41 PM" src="https://github.com/twitte01/232R_GroupProject/assets/168356340/6f3f20bf-43c0-4880-90e5-fce088ff4dd5">

Categorical variables are a common challenge in clustering. One-hot encoding and k-modes clustering are two popular techniques for handling categorical variables. For datasets that contain mixed data, the k-prototypes algorithm and other specialized clustering algorithms can be used.

**Reply by twitte01** [reply]
# Example Code 
```
import numpy as np
import pandas as pd
from sklearn.cluster import KMeans
from kmodes.kmodes import KModes
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from tabulate import tabulate  # Import tabulate

num_samples = 200

# Numerical data representing customer spending behavior
numerical_data = np.random.rand(num_samples, 3) * 1000

# Categorical data representing customer preferences
categorical_data = np.random.choice(['Electronics', 'Clothing', 'HomeDecor'], size=(num_samples, 2))

df = pd.DataFrame(np.concatenate([numerical_data, categorical_data], axis=1),
                  columns=['SpendingScore', 'VisitFrequency', 'AvgPurchaseAmount', 'PreferredCategory1', 'PreferredCategory2'])

print(tabulate(df, headers='keys', tablefmt="pretty"))

# Separate numerical and categorical columns
numerical_columns = ['SpendingScore', 'VisitFrequency', 'AvgPurchaseAmount']
categorical_columns = ['PreferredCategory1', 'PreferredCategory2']

# Create a mixed clustering pipeline
pipeline = Pipeline([
    ('preprocessor', ColumnTransformer(
        transformers=[
            ('num', StandardScaler(), numerical_columns),
            ('cat', OneHotEncoder(), categorical_columns)
        ]
    )),
    ('numerical_clusterer', KMeans(n_clusters=4)),  # Adjusted the number of clusters
    ('categorical_clusterer', KModes(n_clusters=3, init='Huang', n_init=1, verbose=2))
])

# Fit the mixed clustering pipeline to the data
pipeline.fit(df)
```

**Reply by twitte01** [reply]
@PraveenManimaran @vitush99 @CanIGetAnAman @yudon90 May be a better approach than straight K-means. I will work on the one-hot encoding for flowsom
